File: Contribute/Plugin Samples/Ruby/XMLParserPlugin/3rd Party Control/InnerXmlTreeView.rb
require "mscorlib"
require "System, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.IO, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Xml, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Drawing, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Threading, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Collections, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Windows.Forms, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.ComponentModel, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Drawing.Printing, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "SS.Ynote.Engine.Controls, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"

module XMLTreeView
	class InnerXmlTreeView < SSTreeView, IXmlControl # <summary>
		# 
		# </summary>
		def initialize(container)
			@document = nil
			@parent = nil
			@draggedFile = System::String.Empty
			@CurrentNode = nil
			@foundList = ArrayList.new(100)
			@printFont = Font.new("Arial", 10)
			@linesPerPage = 0
			@yPos = 0
			@count = 0
			@indent = System::String.Empty
			@leftMargin = 0
			@topMargin = 0
			@endOfPageReached = false
			@printedNodes = ArrayList.new(1000)
			@eventArgs = nil
			@printDocument1 = PrintDocument.new()
			@parent = container
			AllowDrop = true
			Dock = System.Windows.Forms.DockStyle.Fill
			ImageIndex = -1
			SelectedImageIndex = -1
			Size = System.Drawing.Size.new(150, 130)
			TabIndex = 0
			@document = XmlDocument.new()
			ShowLines = false
		end

		def Search(sender, e)
			if Nodes.Count > 0 then
				SearchDlg.new(self).Show()
			end
		end

		# <summary>
		# 
		# </summary>
		# <param name="criterion"></param>
		# <param name="caseSensitive"></param>
		def StartSearch(criterion, caseSensitive)
			@foundList.Clear()
			self.@caseSensitive = caseSensitive
			@searchText = caseSensitive ? criterion : criterion.ToUpper()
		end

		# <summary>
		# 
		# </summary>
		def Next()
			SelectedNode.BackColor = Color.Empty
			@foundNode = false
			self.RecurseTreeNodes(Nodes)
			if not @foundNode then
				MessageBox.Show("End of tree reached !", XmlTreeView.MessageBoxTitle)
				@foundList.Clear()
			end
		end

		def Print(sender, e)
			if Nodes.Count > 0 and @parent.PrintDialog.ShowDialog() == DialogResult.OK then
				@printedNodes.Clear()
				@printDocument1.Print()
			end
		end

		def PrintPage(sender, e)
			@yPos = 0
			@count = 0
			@endOfPageReached = false
			@eventArgs = e
			@eventArgs.HasMorePages = false
			@leftMargin = @eventArgs.MarginBounds.Left
			@topMargin = @eventArgs.MarginBounds.Top
			@linesPerPage = @eventArgs.MarginBounds.Height / @printFont.GetHeight(@eventArgs.Graphics)
			# Iterate over the file, printing each line.
			self.RecursePrintTreeNodes(Nodes)
		end

		def PrintDocument
			return @printDocument1
		end

		def Edit(sender, e)
			if Nodes.Count > 0 and SelectedNode != nil and @parent.LabelEdit then
				@editingNode = SelectedNode
				if @editingNode.ConnectedXmlElement != nil then
					if not @closingHandlerAssigned then
						parentForm = self.FindForm()
						parentForm.Closing { @parent.form_Closing() }
						@closingHandlerAssigned = true
					end
					height = @editingNode.Bounds.Height
					width = @editingNode.Bounds.Width
					left = @editingNode.Bounds.Left
					top = @editingNode.Bounds.Top
					@editingNode.ExpandAll()
					if @editingNode.ConnectedXmlElement.HasChildNodes and @editingNode.ConnectedXmlElement.FirstChild.NodeType != XmlNodeType.Text then
						height = @editingNode.NextNode.Bounds.Bottom - @editingNode.Bounds.Top
						width = Width - left
					end
					@editBox = TextBox.new()
					@editBox.Multiline = true
					@editBox.BorderStyle = BorderStyle.FixedSingle
					@editBox.Leave { |sender, e| self.editBox_Leave(sender, e) }
					@editBox.KeyUp { |sender, e| self.editBox_KeyUp(sender, e) }
					@editBox.SetBounds(left, top, width, height)
					@editingNode.RecurseSubNodes(@editingNode.Parent)
					@editBox.Text = @editingNode.SelfAndChildren
					Controls.Add(@editBox)
					@editBox.Focus()
				end
			end
		end

		def Copy(sender, e)
			if Nodes.Count > 0 then
				Clipboard.SetDataObject(Xml, true)
			end
		end

		def Paste(sender, e)
			dataObject = Clipboard.GetDataObject()
			if dataObject.GetDataPresent(DataFormats.Text) then
				@parent.Xml = dataObject.GetData(DataFormats.Text)
			end
		end

		def Delete(sender, e)
			if Nodes.Count > 0 and nil != SelectedNode and @parent.LabelEdit then
				tmp = Xml
				begin
					elemToRemove = (SelectedNode).ConnectedXmlElement
					if nil != elemToRemove then
						elemToRemove.ParentNode.RemoveChild(elemToRemove)
						if SelectedNode.NextNode.Text.Equals("</" + elemToRemove.Name + ">") then
							SelectedNode.NextNode.Remove()
						end
						SelectedNode.Remove()
					end
				rescue  => 
					MessageBox.Show("Cannot delete this node ! Rolling back...", XmlTreeView.MessageBoxTitle)
					@parent.Xml = tmp
				ensure
				end
			end
		end

		def Save(sender, e)
			if Nodes.Count > 0 and @parent.SaveDialog.ShowDialog() == DialogResult.OK then
				sw = StreamWriter.new(@parent.SaveDialog.FileName)
				sw.Write(Xml)
				@parent.fileXml = Xml
				@parent.filePath = @parent.SaveDialog.FileName
			end
		end

		# <summary>
		# Takes XML-data as a value.
		# </summary>
		def Xml
			return @document.OuterXml
		end

		def Xml=(value)
			if value.Length > 0 then
				self.LoadXml(value)
			end
		end

		def editBox_Leave(sender, e)
			if Controls.Contains(@editBox) then
				@editBox.Leave.remove(OnEventHandler.new(editBox_Leave))
				@editBox.KeyUp.remove(OnKeyEventHandler.new(editBox_KeyUp))
				Controls.Remove(@editBox)
				@editingNode.Text = @editBox.Text
				@editBox.Dispose()
			end
		end

		def editBox_KeyUp(sender, e)
			if e.KeyCode == Keys.Escape then
				@editBox.Leave.remove(OnEventHandler.new(editBox_Leave))
				@editBox.KeyUp.remove(OnKeyEventHandler.new(editBox_KeyUp))
				Controls.Remove(@editBox)
				@editBox.Dispose()
			end
		end

		def RecursePrintTreeNodes(coll)
			if not @endOfPageReached then
				enumerator = coll.GetEnumerator()
				while enumerator.MoveNext()
					node = enumerator.Current
					if @endOfPageReached then
						break
					end
					if not @printedNodes.Contains(node) then
						textToDraw = @indent + node.Text
						textWidthPx = 0
						@yPos = @topMargin + (@count * @printFont.GetHeight())
						if (textWidthPx = @eventArgs.Graphics.MeasureString(textToDraw, @printFont).Width) > @eventArgs.MarginBounds.Width then
							startPos = 0
							pixPerChar = textWidthPx / textToDraw.Length
							maxCharsPerLine = (@eventArgs.MarginBounds.Width / pixPerChar)
							while (startPos + maxCharsPerLine) < textToDraw.Length
								@eventArgs.Graphics.DrawString(textToDraw.Substring(startPos, maxCharsPerLine), @printFont, Brushes.Black, @leftMargin, @yPos, StringFormat.new())
								startPos += maxCharsPerLine
								@yPos += @printFont.GetHeight()
								@count += 1
							end
							@eventArgs.Graphics.DrawString(textToDraw.Substring(startPos), @printFont, Brushes.Black, @leftMargin, @yPos, StringFormat.new())
						else
							@eventArgs.Graphics.DrawString(textToDraw, @printFont, Brushes.Black, @leftMargin, @yPos, StringFormat.new())
						end
						@count += 1
						@printedNodes.Add(node)
					end
					if @endOfPageReached = (@count >= @linesPerPage) then
						@eventArgs.HasMorePages = true
						break
					end
					if node.Nodes.Count > 0 then
						@indent += "    "
						self.RecursePrintTreeNodes(node.Nodes)
					end
				end
			end
			if @indent.Length > 0 then
				@indent = @indent.Substring(0, @indent.Length - 4)
			end
		end

		def RecurseTreeNodes(nodes)
			if not @foundNode then
				nodeText = nil
				enumerator = nodes.GetEnumerator()
				while enumerator.MoveNext()
					node = enumerator.Current
					if @foundNode then
						break
					end
					nodeText = @caseSensitive ? node.Text : node.Text.ToUpper()
					if nodeText.IndexOf(@searchText) > -1 and not @foundList.Contains(node) then
						SelectedNode = node
						SelectedNode.BackColor = Color.Blue
						@foundList.Add(node)
						@foundNode = true
						break
					end
					if node.Nodes.Count > 0 then
						self.RecurseTreeNodes(node.Nodes)
					end
				end
			end
		end

		def LoadXml(xml)
			self.SuspendLayout()
			@document.LoadXml(xml)
			Nodes.Clear()
			if xml.StartsWith("<?") then
				Nodes.Add(XmlTreeNode.new(xml.Substring(0, xml.IndexOf("?>") + 2), nil))
			end
			self.RecurseAndAssignNodes(@document.DocumentElement)
			self.ExpandAll()
			self.ResumeLayout(false)
		end

		def RecurseAndAssignNodes(elem)
			attrs = System::String.Empty
			addedNode = nil
			if elem.NodeType == XmlNodeType.Element then
				enumerator = elem.Attributes.GetEnumerator()
				while enumerator.MoveNext()
					attr = enumerator.Current
					attrs += " " + attr.Name + "=\"" + attr.Value + "\""
				end
			end
			if elem.Equals(@document.DocumentElement) then
				addedNode = XmlTreeNode.new("<" + elem.Name + attrs + ">", elem)
				Nodes.Add(addedNode)
				InnerXmlTreeView.CurrentNode = addedNode
				Nodes.Add(XmlTreeNode.new("</" + elem.Name + ">", nil))
			elsif elem.HasChildNodes and elem.ChildNodes[0].NodeType == XmlNodeType.Text then
				addedNode = XmlTreeNode.new("<" + elem.Name + attrs + ">" + elem.InnerText + "</" + elem.Name + ">", elem)
				InnerXmlTreeView.CurrentNode.Nodes.Add(addedNode)
				InnerXmlTreeView.CurrentNode = addedNode
			elsif elem and (elem).IsEmpty then
				addedNode = XmlTreeNode.new("<" + elem.Name + attrs + "/>", elem)
				InnerXmlTreeView.CurrentNode.Nodes.Add(addedNode)
				InnerXmlTreeView.CurrentNode = addedNode
			else
				addedNode = XmlTreeNode.new("<" + elem.Name + attrs + ">", elem)
				InnerXmlTreeView.CurrentNode.Nodes.Add(addedNode)
				InnerXmlTreeView.CurrentNode = addedNode
				InnerXmlTreeView.CurrentNode.Parent.Nodes.Add(XmlTreeNode.new("</" + elem.Name + ">", nil))
			end
			enumerator = elem.ChildNodes.GetEnumerator()
			while enumerator.MoveNext()
				child = enumerator.Current
				if child.NodeType == XmlNodeType.Element then
					self.RecurseAndAssignNodes(child)
				elsif child.NodeType == XmlNodeType.Comment then
					InnerXmlTreeView.CurrentNode.Nodes.Add(XmlTreeNode.new(child.OuterXml, child))
				end
			end
			if InnerXmlTreeView.CurrentNode.Parent != nil then
				InnerXmlTreeView.CurrentNode = InnerXmlTreeView.CurrentNode.Parent
			end
		end

		def OnKeyUp(e)
			self.OnKeyUp(e)
			if e.Control and e.KeyCode == Keys.C then
				self.Copy(nil, nil)
			elsif e.Control and e.KeyCode == Keys.V then
				self.Paste(nil, nil)
			elsif e.Control and e.KeyCode == Keys.F then
				self.Search(nil, nil)
			elsif e.Control and e.KeyCode == Keys.P then
				self.Print(nil, nil)
			elsif e.Control and e.KeyCode == Keys.S then
				self.Save(nil, nil)
			elsif e.KeyCode == Keys.Delete then
				self.Delete(nil, nil)
			elsif e.KeyCode == Keys.Insert then
				self.Edit(nil, nil)
			end
		end

		def OnDragEnter(e)
			self.OnDragEnter(e)
			if e.Data.GetDataPresent(DataFormats.FileDrop) then
				e.Effect = DragDropEffects.Copy
				@draggedFile = (e.Data.GetData(DataFormats.FileDrop))[0]
			end
		end

		def OnDragDrop(e)
			self.OnDragDrop(e)
			if @draggedFile.Length > 0 then
				fs = StreamReader.new(@draggedFile)
				@parent.fileXml = fs.ReadToEnd()
				@parent.Xml = @parent.fileXml
				@parent.filePath = @draggedFile
				@draggedFile = System::String.Empty
			end
		end

		def OnDoubleClick(e)
			if SelectedNode.Parent != nil and (SelectedNode).ConnectedXmlElement != nil then
				self.Edit(nil, nil)
			else
			end
		end
	end
end

File: Contribute/Plugin Samples/IronPython/XMLParserPlugin/3rd Party Control/InnerXmlTreeView.py
from System import *
from System.IO import *
from System.Xml import *
from System.Drawing import *
from System.Threading import *
from System.Collections import *
from System.Windows.Forms import *
from System.ComponentModel import *
from System.Drawing.Printing import *
from SS.Ynote.Engine.Controls import *

class InnerXmlTreeView(SSTreeView, IXmlControl):
	def __init__(self, container):
		""" <summary>
		 
		 </summary>
		"""
		self._document = None
		self._parent = None
		self._draggedFile = str.Empty
		self._CurrentNode = None
		self._foundList = ArrayList(100)
		self._printFont = Font("Arial", 10)
		self._linesPerPage = 0
		self._yPos = 0
		self._count = 0
		self._indent = str.Empty
		self._leftMargin = 0
		self._topMargin = 0
		self._endOfPageReached = False
		self._printedNodes = ArrayList(1000)
		self._eventArgs = None
		self._printDocument1 = PrintDocument()
		self._parent = container
		AllowDrop = True
		Dock = System.Windows.Forms.DockStyle.Fill
		ImageIndex = -1
		SelectedImageIndex = -1
		Size = System.Drawing.Size(150, 130)
		TabIndex = 0
		self._document = XmlDocument()
		ShowLines = False

	def Search(self, sender, e):
		if Nodes.Count > 0:
			SearchDlg(self).Show()

	def StartSearch(self, criterion, caseSensitive):
		""" <summary>
		 
		 </summary>
		 <param name="criterion"></param>
		 <param name="caseSensitive"></param>
		"""
		self._foundList.Clear()
		self._caseSensitive = caseSensitive
		self._searchText = criterion if caseSensitive else criterion.ToUpper()

	def Next(self):
		""" <summary>
		 
		 </summary>
		"""
		SelectedNode.BackColor = Color.Empty
		self._foundNode = False
		self.RecurseTreeNodes(Nodes)
		if not self._foundNode:
			MessageBox.Show("End of tree reached !", XmlTreeView.MessageBoxTitle)
			self._foundList.Clear()

	def Print(self, sender, e):
		if Nodes.Count > 0 and self._parent.PrintDialog.ShowDialog() == DialogResult.OK:
			self._printedNodes.Clear()
			self._printDocument1.Print()

	def PrintPage(self, sender, e):
		self._yPos = 0
		self._count = 0
		self._endOfPageReached = False
		self._eventArgs = e
		self._eventArgs.HasMorePages = False
		self._leftMargin = self._eventArgs.MarginBounds.Left
		self._topMargin = self._eventArgs.MarginBounds.Top
		self._linesPerPage = self._eventArgs.MarginBounds.Height / self._printFont.GetHeight(self._eventArgs.Graphics)
		# Iterate over the file, printing each line.
		self.RecursePrintTreeNodes(Nodes)

	def get_PrintDocument(self):
		return self._printDocument1

	PrintDocument = property(fget=get_PrintDocument)

	def Edit(self, sender, e):
		if Nodes.Count > 0 and SelectedNode != None and self._parent.LabelEdit:
			self._editingNode = SelectedNode
			if self._editingNode.ConnectedXmlElement != None:
				if not self._closingHandlerAssigned:
					parentForm = self.FindForm()
					parentForm.Closing += self._parent.form_Closing
					self._closingHandlerAssigned = True
				height = self._editingNode.Bounds.Height
				width = self._editingNode.Bounds.Width
				left = self._editingNode.Bounds.Left
				top = self._editingNode.Bounds.Top
				self._editingNode.ExpandAll()
				if self._editingNode.ConnectedXmlElement.HasChildNodes and self._editingNode.ConnectedXmlElement.FirstChild.NodeType != XmlNodeType.Text:
					height = self._editingNode.NextNode.Bounds.Bottom - self._editingNode.Bounds.Top
					width = Width - left
				self._editBox = TextBox()
				self._editBox.Multiline = True
				self._editBox.BorderStyle = BorderStyle.FixedSingle
				self._editBox.Leave += self.editBox_Leave
				self._editBox.KeyUp += self.editBox_KeyUp
				self._editBox.SetBounds(left, top, width, height)
				self._editingNode.RecurseSubNodes(self._editingNode.Parent)
				self._editBox.Text = self._editingNode.SelfAndChildren
				Controls.Add(self._editBox)
				self._editBox.Focus()

	def Copy(self, sender, e):
		if Nodes.Count > 0:
			Clipboard.SetDataObject(Xml, True)

	def Paste(self, sender, e):
		dataObject = Clipboard.GetDataObject()
		if dataObject.GetDataPresent(DataFormats.Text):
			self._parent.Xml = dataObject.GetData(DataFormats.Text)

	def Delete(self, sender, e):
		if Nodes.Count > 0 and None != SelectedNode and self._parent.LabelEdit:
			tmp = Xml
			try:
				elemToRemove = (SelectedNode).ConnectedXmlElement
				if None != elemToRemove:
					elemToRemove.ParentNode.RemoveChild(elemToRemove)
					if SelectedNode.NextNode.Text.Equals("</" + elemToRemove.Name + ">"):
						SelectedNode.NextNode.Remove()
					SelectedNode.Remove()
			except , :
				MessageBox.Show("Cannot delete this node ! Rolling back...", XmlTreeView.MessageBoxTitle)
				self._parent.Xml = tmp
			finally:

	def Save(self, sender, e):
		if Nodes.Count > 0 and self._parent.SaveDialog.ShowDialog() == DialogResult.OK:

	# <summary>
	# Takes XML-data as a value.
	# </summary>
	def get_Xml(self):
		return self._document.OuterXml

	def set_Xml(self, value):
		if value.Length > 0:
			self.LoadXml(value)

	Xml = property(fget=get_Xml, fset=set_Xml)

	def editBox_Leave(self, sender, e):
		if Controls.Contains(self._editBox):
			self._editBox.Leave -= self.editBox_Leave
			self._editBox.KeyUp -= self.editBox_KeyUp
			Controls.Remove(self._editBox)
			self._editingNode.Text = self._editBox.Text
			self._editBox.Dispose()

	def editBox_KeyUp(self, sender, e):
		if e.KeyCode == Keys.Escape:
			self._editBox.Leave -= self.editBox_Leave
			self._editBox.KeyUp -= self.editBox_KeyUp
			Controls.Remove(self._editBox)
			self._editBox.Dispose()

	def RecursePrintTreeNodes(self, coll):
		if not self._endOfPageReached:
			enumerator = coll.GetEnumerator()
			while enumerator.MoveNext():
				node = enumerator.Current
				if self._endOfPageReached:
					break
				if not self._printedNodes.Contains(node):
					textToDraw = self._indent + node.Text
					textWidthPx = 0
					self._yPos = self._topMargin + (self._count * self._printFont.GetHeight())
					if (textWidthPx = self._eventArgs.Graphics.MeasureString(textToDraw, self._printFont).Width) > self._eventArgs.MarginBounds.Width:
						startPos = 0
						pixPerChar = textWidthPx / textToDraw.Length
						maxCharsPerLine = (self._eventArgs.MarginBounds.Width / pixPerChar)
						while (startPos + maxCharsPerLine) < textToDraw.Length:
							self._eventArgs.Graphics.DrawString(textToDraw.Substring(startPos, maxCharsPerLine), self._printFont, Brushes.Black, self._leftMargin, self._yPos, StringFormat())
							startPos += maxCharsPerLine
							self._yPos += self._printFont.GetHeight()
							self._count += 1
						self._eventArgs.Graphics.DrawString(textToDraw.Substring(startPos), self._printFont, Brushes.Black, self._leftMargin, self._yPos, StringFormat())
					else:
						self._eventArgs.Graphics.DrawString(textToDraw, self._printFont, Brushes.Black, self._leftMargin, self._yPos, StringFormat())
					self._count += 1
					self._printedNodes.Add(node)
				if self._endOfPageReached = (self._count >= self._linesPerPage):
					self._eventArgs.HasMorePages = True
					break
				if node.Nodes.Count > 0:
					self._indent += "    "
					self.RecursePrintTreeNodes(node.Nodes)
		if self._indent.Length > 0:
			self._indent = self._indent.Substring(0, self._indent.Length - 4)

	def RecurseTreeNodes(self, nodes):
		if not self._foundNode:
			nodeText = None
			enumerator = nodes.GetEnumerator()
			while enumerator.MoveNext():
				node = enumerator.Current
				if self._foundNode:
					break
				nodeText = node.Text if self._caseSensitive else node.Text.ToUpper()
				if nodeText.IndexOf(self._searchText) > -1 and not self._foundList.Contains(node):
					SelectedNode = node
					SelectedNode.BackColor = Color.Blue
					self._foundList.Add(node)
					self._foundNode = True
					break
				if node.Nodes.Count > 0:
					self.RecurseTreeNodes(node.Nodes)

	def LoadXml(self, xml):
		self.SuspendLayout()
		self._document.LoadXml(xml)
		Nodes.Clear()
		if xml.StartsWith("<?"):
			Nodes.Add(XmlTreeNode(xml.Substring(0, xml.IndexOf("?>") + 2), None))
		self.RecurseAndAssignNodes(self._document.DocumentElement)
		self.ExpandAll()
		self.ResumeLayout(False)

	def RecurseAndAssignNodes(self, elem):
		attrs = str.Empty
		addedNode = None
		if elem.NodeType == XmlNodeType.Element:
			enumerator = elem.Attributes.GetEnumerator()
			while enumerator.MoveNext():
				attr = enumerator.Current
				attrs += " " + attr.Name + "=\"" + attr.Value + "\""
		if elem.Equals(self._document.DocumentElement):
			addedNode = XmlTreeNode("<" + elem.Name + attrs + ">", elem)
			Nodes.Add(addedNode)
			InnerXmlTreeView.CurrentNode = addedNode
			Nodes.Add(XmlTreeNode("</" + elem.Name + ">", None))
		elif elem.HasChildNodes and elem.ChildNodes[0].NodeType == XmlNodeType.Text:
			addedNode = XmlTreeNode("<" + elem.Name + attrs + ">" + elem.InnerText + "</" + elem.Name + ">", elem)
			InnerXmlTreeView.CurrentNode.Nodes.Add(addedNode)
			InnerXmlTreeView.CurrentNode = addedNode
		elif  and (elem).IsEmpty:
			addedNode = XmlTreeNode("<" + elem.Name + attrs + "/>", elem)
			InnerXmlTreeView.CurrentNode.Nodes.Add(addedNode)
			InnerXmlTreeView.CurrentNode = addedNode
		else:
			addedNode = XmlTreeNode("<" + elem.Name + attrs + ">", elem)
			InnerXmlTreeView.CurrentNode.Nodes.Add(addedNode)
			InnerXmlTreeView.CurrentNode = addedNode
			InnerXmlTreeView.CurrentNode.Parent.Nodes.Add(XmlTreeNode("</" + elem.Name + ">", None))
		enumerator = elem.ChildNodes.GetEnumerator()
		while enumerator.MoveNext():
			child = enumerator.Current
			if child.NodeType == XmlNodeType.Element:
				self.RecurseAndAssignNodes(child)
			elif child.NodeType == XmlNodeType.Comment:
				InnerXmlTreeView.CurrentNode.Nodes.Add(XmlTreeNode(child.OuterXml, child))
		if InnerXmlTreeView.CurrentNode.Parent != None:
			InnerXmlTreeView.CurrentNode = InnerXmlTreeView.CurrentNode.Parent

	def OnKeyUp(self, e):
		self.OnKeyUp(e)
		if e.Control and e.KeyCode == Keys.C:
			self.Copy(None, None)
		elif e.Control and e.KeyCode == Keys.V:
			self.Paste(None, None)
		elif e.Control and e.KeyCode == Keys.F:
			self.Search(None, None)
		elif e.Control and e.KeyCode == Keys.P:
			self.Print(None, None)
		elif e.Control and e.KeyCode == Keys.S:
			self.Save(None, None)
		elif e.KeyCode == Keys.Delete:
			self.Delete(None, None)
		elif e.KeyCode == Keys.Insert:
			self.Edit(None, None)

	def OnDragEnter(self, e):
		self.OnDragEnter(e)
		if e.Data.GetDataPresent(DataFormats.FileDrop):
			e.Effect = DragDropEffects.Copy
			self._draggedFile = (e.Data.GetData(DataFormats.FileDrop))[0]

	def OnDragDrop(self, e):
		self.OnDragDrop(e)
		if self._draggedFile.Length > 0:
			self._draggedFile = str.Empty

	def OnDoubleClick(self, e):
		if SelectedNode.Parent != None and (SelectedNode).ConnectedXmlElement != None:
			self.Edit(None, None)
		else: